zebra: handle renamed interfaces at netlink level

pguibert6WIND opened the pull request:
when an interface is renamed in a given namespace, the ipv6
addresses are maintained in the new interface. The fix consists
in using the same interface context, instead of creating a new
one.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
https://github.com/FRRouting/frr/issues/8256

qlyoung commented:
@polychaeta rereview

qlyoung commented:
@polychaeta rereview

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/1e285fcedee7a15a19bc8d83269ad6cf/raw/a0ab94368a891b2ec4f7626e3b4840670b205e8c/cr_8310_1616446940.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/if_netlink.c b/zebra/if_netlink.c
index a7e32397d..3b3961d6a 100644
--- a/zebra/if_netlink.c
+++ b/zebra/if_netlink.c
@@ -107,14 +107,13 @@ static void set_if_change(struct interface *ifp, ifindex_t ifi_index,
 	if (((oifp = if_lookup_by_name_per_ns(zns, if_name)) != NULL)
 	    && (oifp != ifp)) {
 		if (IS_ZEBRA_DEBUG_KERNEL)
-			zlog_debug(
-				   "interface name %s renamed to %s",
+			zlog_debug("interface name %s renamed to %s",
 				   oifp->name, ifp->name);
 		if (if_is_up(oifp))
 			flog_err(
-				 EC_LIB_INTERFACE,
-				 "interface rename detected on up interface from %s to %s, results are uncertain!",
-				 oifp->name, ifp->name);
+				EC_LIB_INTERFACE,
+				"interface rename detected on up interface from %s to %s, results are uncertain!",
+				oifp->name, ifp->name);
 		update = true;
 	}
 	if_set_name(ifp, if_name);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

vjardin commented:
LGTM,

louis-oui commented:
Hello @vjardin, could you merge this PR please ?

idryzhov commented on zebra/if_netlink.c:
```diff
@@ -98,10 +99,26 @@ static void set_ifindex(struct interface *ifp, ifindex_t ifi_index,
 	}
 	if_set_index(ifp, ifi_index);
+
+	oifp = if_lookup_by_name_per_ns(zns, if_name);
```
Here you rewrite the oifp pointer that was obtained by `if_lookup_by_index_per_ns` earlier.

pguibert6WIND replied:
this change of code was useless. youre right.

pguibert6WIND replied:
@idryzhov , is it ok for you ?

idryzhov commented on zebra/if_netlink.c:
```diff
@@ -98,10 +99,26 @@ static void set_ifindex(struct interface *ifp, ifindex_t ifi_index,
 					"interface rename detected on up interface: index %d was renamed from %s to %s, results are uncertain!",
 					ifi_index, oifp->name, ifp->name);
-			if_delete_update(oifp);
+			update = true;
```
If we don't do `if_delete_update` here, then the call to `if_set_index` a couple of lines later won't work as expected, because of the ifindex collision.

idryzhov reviewed (changes requested):
I left a couple of comments, but I also don't understand what is wrong with the current code.

Here is the current code flow in `netlink_link_change` when the interface is renamed:
- `ifp = if_lookup_by_name_per_ns(zns, name);`
Lookup for an existing interface by name. It either returns NULL or a pointer to an existing inactive interface with an unknown ifindex.
- `if (ifp == NULL) ifp = if_get_by_name(name, vrf_id);`
If there's no inactive interface found on the first step, the new one is created.
- `set_ifindex(ifp, ifi->ifi_index, zns);`
  * First, this function tries to find an existing interface with the same ifindex. If this is a rename, then `oifp` is a pointer to an old interface.
  * Then it makes it inactive (resetting the ifindex) by calling `if_delete_update`.
  * And then it assigns the ifindex to a newly created interface by calling to `if_set_index`.

The current flow looks completely correct to me. Please, explain what is wrong with it.

pguibert6WIND commented:
I kept only the code that searches for original interface .

pguibert6WIND commented:
@idryzhov , I simplfied the code change.
are you okay with that change?

idryzhov reviewed (changes requested):
I still don't think it's the right fix.

You don't change the code path for the case when we already have a config for the new name (`!CHECK_FLAG(ifp->status, ZEBRA_INTERFACE_ACTIVE)`). So whatever you're trying to fix won't be fixed in this case.
For the case when we don't have a config for the new name, you don't update the name of the old interface object. This is also not correct.

Overall this change makes the code behavior inconsistent. Sometimes we create a new object for a new name and sometimes we don't. The current code is consistent and always creates a new object.

pguibert6WIND commented:
> I still don't think it's the right fix.
> 
> You don't change the code path for the case when we already have a config for the new name (`!CHECK_FLAG(ifp->status, ZEBRA_INTERFACE_ACTIVE)`). So whatever you're trying to fix won't be fixed in this case.
> For the case when we don't have a config for the new name, you don't update the name of the old interface object. This is also not correct.
> 
> Overall this change makes the code behavior inconsistent. Sometimes we create a new object for a new name and sometimes we don't. The current code is consistent and always creates a new object.

sorry for delay. I was away from keyboard for a few days.
I tried to address the issue of the code path change in nb api.
I try to reuse some code from https://github.com/FRRouting/frr/pull/8864/commits/763725cd5e431cb4f4ec385e35b312cc7807163a

lets continue discussion on this ticket.

idryzhov commented on lib/if.c:
```diff
@@ -231,6 +231,43 @@ struct interface *if_create_ifindex(ifindex_t ifindex, vrf_id_t vrf_id)
+		snprintf(newpath, sizeof(newpath),
+			 "/frr-interface:lib/interface[name='%s'][vrf='%s']",
+			 name, vrf->name);
+		if_dnode = yang_dnode_getf(running_config->dnode, "%s/vrf",
```
You're not changing the name here, you're changing the VRF. More than that, oldpath and newpath are the same because you're using `ifp->name` and `name` to distinguish and they are the same after the call to `if_set_name` at the beginning of the function. Is this code tested?

pguibert6WIND replied:
I tested it like below:
1. vtysh
```
config terminal
password zebra
interface dum3
ip address 4.4.4.4/24
end
```
2. check yang context: ( telnet 0 2601, then enable)
```
#show yang operational-context /frr-interface:lib
```
3. shell creation
```
ip link add dum3 type dummy
ip link set dev dum3 up
```
4. interface rename
```
ip link set dev dum3 down
ip link set dev dum3 name dum4
```
5. check: show yang operational
```
dut-vm# show yang operational-data /frr-interface:lib
{
  "frr-interface:lib": {
    "interface": [
      { 
        "name": "dum4",
        "vrf": "default",
        "state": {
          "if-index": 6,
          "mtu": 1500,
          "mtu6": 1500,
          "speed": 0,
          "metric": 0,
          "phy-address": "5e:37:28:23:c0:d6"
        },
        "frr-zebra:zebra": {
          "state": {
            "up-count": 1,
            "down-count": 1,
            "ptm-status": "disabled"
          }
        }
      },
```

I agree with the code done, there seems to be something wrong. however, I can see the operation context ok.
is there something else I should use ?

idryzhov replied:
Operational context is just a representation of the daemon's internal state.
You should use `show conf run json zebra` (or other daemon) to check the daemon's NB config.

pguibert6WIND replied:
updated with appropriate code ( modifying interface name instead of vrf name).
and checked that with traces I enter in the change_leaf code.

idryzhov reviewed (changes requested):
Given more thought about this, I don't think that we should handle the interface rename any different from how it's done now. From the FRR perspective, the interface rename is the deletion of the old interface and the addition of a new one. We may already have a config for the new name – and in this case, we apply this config. If we don't have a config then it's just a new clean interface and this is also correct.
We can not change NB config in case of rename – the config must not depend on any operational changes.

pguibert6WIND commented:
> Given more thought about this, I don't think that we should handle the interface rename any different from how it's done now. From the FRR perspective, the interface rename is the deletion of the old interface and the addition of a new one. We may already have a config for the new name – and in this case, we apply this config. If we don't have a config then it's just a new clean interface and this is also correct.
> We can not change NB config in case of rename – the config must not depend on any operational changes.

This is where I try to fix the ipv6 address lost on frr config.
On this situation If I continue removing old interface, on zebra, then the new interface is not aware of the presence of the ipv6 address. This is what pushes me to use rename.

What also pushes me to use rename is that we know we have problems with "rename" ( see https://github.com/opensourcerouting/frr/wiki/Advanced-topics#configuration-changes-coming-from-the-kernel ).

If you say we should continue create a new copy, then you should explain me how to resync the kernel config with the new interface config. If you dont know, I would like to share this with a second pair of eyes.

idryzhov commented:
I definitely agree we should discuss this with a wider audience because we're going in circles here :)
Could you, please, explain a little bit more about the config being lost? Is there are IPv6 addresses configured outside of FRR and zebra doesn't add them to the new interface context? Or are there addresses configured inside FRR for the new interface name and this config is not being applied?

pguibert6WIND commented:
> I definitely agree we should discuss this with a wider audience because we're going in circles here :)
> Could you, please, explain a little bit more about the config being lost? Are there are IPv6 addresses configured outside of FRR and zebra doesn't add them to the new interface context? Or are there addresses configured inside FRR for the new interface name and this config is not being applied?

I want to solve issue described in https://github.com/FRRouting/frr/issues/8256.
This is not a FRR configuration issue. IPv6 config takes place outside of zebra.
IPv6 global addresses are not retransmitted to zebra once interface is renamed and moved to up.
The address was on the interface that has been deleted, and should be somehow copied to the new interface context.

pguibert6WIND commented:
ci:rerun

idryzhov commented:
> > I definitely agree we should discuss this with a wider audience because we're going in circles here :)
> > Could you, please, explain a little bit more about the config being lost? Are there are IPv6 addresses configured outside of FRR and zebra doesn't add them to the new interface context? Or are there addresses configured inside FRR for the new interface name and this config is not being applied?
> 
> I want to solve issue described in #8256.
> This is not a FRR configuration issue. IPv6 config takes place outside of zebra.
> IPv6 global addresses are not retransmitted to zebra once interface is renamed and moved to up.
> The address was on the interface that has been deleted, and should be somehow copied to the new interface context.

Ok, I understand the problem, but this is definitely not the right way to solve it.
The code you wrote will work differently when we have or don't have the config for the new name:
- when we don't have the config, your code fixes the issue by preserving the interface context
- but when we have the config - the issue stays because nothing is changed

You should find the real reason why we don't receive the information about IPv6 addresses for the new name.
As you wrote in the linked issue – IPv4 works correctly, so this shouldn't be the overall issue, right? This should be something related only to IPv6. We can discuss this on a Tuesday meeting if you want.

pguibert6WIND commented:
> > > I definitely agree we should discuss this with a wider audience because we're going in circles here :)
> > > Could you, please, explain a little bit more about the config being lost? Are there are IPv6 addresses configured outside of FRR and zebra doesn't add them to the new interface context? Or are there addresses configured inside FRR for the new interface name and this config is not being applied?
> > 
> > 
> > I want to solve issue described in #8256.
> > This is not a FRR configuration issue. IPv6 config takes place outside of zebra.
> > IPv6 global addresses are not retransmitted to zebra once interface is renamed and moved to up.
> > The address was on the interface that has been deleted, and should be somehow copied to the new interface context.
> 
> Ok, I understand the problem, but this is definitely not the right way to solve it.
> The code you wrote will work differently when we have or don't have the config for the new name:
> 
>     * when we don't have the config, your code fixes the issue by preserving the interface context
> 
>     * but when we have the config - the issue stays because nothing is changed
> 
> 
> You should find the real reason why we don't receive the information about IPv6 addresses for the new name.
> As you wrote in the linked issue – IPv4 works correctly, so this shouldn't be the overall issue, right? This should be something related only to IPv6. We can discuss this on a Tuesday meeting if you want.

note that I fixes the

> > > I definitely agree we should discuss this with a wider audience because we're going in circles here :)
> > > Could you, please, explain a little bit more about the config being lost? Are there are IPv6 addresses configured outside of FRR and zebra doesn't add them to the new interface context? Or are there addresses configured inside FRR for the new interface name and this config is not being applied?
> > 
> > 
> > I want to solve issue described in #8256.
> > This is not a FRR configuration issue. IPv6 config takes place outside of zebra.
> > IPv6 global addresses are not retransmitted to zebra once interface is renamed and moved to up.
> > The address was on the interface that has been deleted, and should be somehow copied to the new interface context.
> 
> Ok, I understand the problem, but this is definitely not the right way to solve it.
> The code you wrote will work differently when we have or don't have the config for the new name:
> 
>     * when we don't have the config, your code fixes the issue by preserving the interface context
> 
>     * but when we have the config - the issue stays because nothing is changed
> 
> 
> You should find the real reason why we don't receive the information about IPv6 addresses for the new name.
> As you wrote in the linked issue – IPv4 works correctly, so this shouldn't be the overall issue, right? This should be something related only to IPv6. We can discuss this on a Tuesday meeting if you want.

ok, I managed to run the test and use 'show conf run json zebra', and I could see that the configuration aligns with the new interface name.
```
dut-vm# show conf run json zebra
{
  "frr-interface:lib": {
    "interface": [
      {
        "name": "dum4",
        "vrf": "default",
        "description": "bla bla bla"
      }
    ]
  }
}
```
From netlink perspective, the ipv6 address initially set on interface is kept on the new interface (thanks to keep_addr_on_down settings).
The problem that is being solved, then, is just a configuration consistency, and what is achieved is that, by renaming config old interface to config new interface, the information stored in the yang context ( description) is kept across the new interface.

idryzhov reviewed (changes requested):
The code doesn't work correctly when we already have the config for the new name:
```
zebra[25679]: [J1C1S-S285Q][EC 100663305] if_set_name(after): corruption detected -- interface with this name exists already in VRF 0!
```

We end up having two interface contexts with the same interface name.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/dc70002904ec658203ead53bcc63dc14/raw/6d3cc0fc2fb25ffb06ee33c53ea202f5c039631d/cr_8310_1626697058.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/if_netlink.c b/zebra/if_netlink.c
index 1a3279168..c72cbf006 100644
--- a/zebra/if_netlink.c
+++ b/zebra/if_netlink.c
@@ -1598,13 +1598,15 @@ int netlink_link_change(struct nlmsghdr *h, ns_id_t ns_id, int startup)
 					if (strcmp(ifp->name, name)) {
 						if (IS_ZEBRA_DEBUG_KERNEL)
 							zlog_debug(
-								   "interface index %d was renamed from %s to %s",
-								   ifp->ifindex, ifp->name,
-								   name);
-						if_update_to_new_name(ifp, name);
+								"interface index %d was renamed from %s to %s",
+								ifp->ifindex,
+								ifp->name,
+								name);
+						if_update_to_new_name(ifp,
+								      name);
 					}
-					/* else interface with same name already present
-					 * reuse ifp pointer
+					/* else interface with same name already
+					 * present reuse ifp pointer
 					 */
 				} else
 					/* unknown interface */

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

pguibert6WIND commented:
with last commit, there should not be corruption in list:
```
ifp = if_lookup_by_name(name)
if (ifp == NULL || !CHECK_FLAG(ifp->status, ACTIVE)) {
    if (ifp == NULL) {
      ifp = if_lookup_by_index(ifi->ifi_index)
      if (ifp) {
          if (strcmp(ifp->name, name)) {
             if_update_to_new_name(ifp, name)                                 <--- if rename, then move NB config if name
          }
      } else {
         ifp = if_get_by_name(name)                                               <--- ifp not found in config, allocate it
     }
 } else {
      /* pre-configured interface, iface already present */                     <--- if iface with name already present, then reuse it
                                                                                                           <--- if vrf changed, move NB config vrf name
}
```

idryzhov commented:
The log I provided is from the actual testing of your latest code. `if_lookup_by_index` doesn't return you the existing fake-interface context and you corrupt the list when you call `if_update_to_new_name`.

I am really concerned that I have to spend so much time on this. Please, thoroughly test your code for both situations – when we have and don't have the config for the new name. The first case doesn't work right now.

donaldsharp commented:
Closing PR hasn't been touched in 2 years
